### PR TITLE
Use a static notification function for PQsetNoticeReceiver. Fixes #300

### LINKED ISSUE #300: Segfault in the psycopg python only variant.
Hi there, I am trying to get pyscopg working on Django (https://github.com/django/django/pull/15687/files), thank you for your work so far on this -- I'll try to get it over the finish line.

One thing I did note though is that if I run the testsuite like this (this is Django's testsuite from my branch with psycopg3 installed):
```
./runtests.py --settings=test_postgresql --parallel=8
cat test_postgresql.py 
# This is an example test settings file for use with the Django test suite.
#
# The 'sqlite3' backend requires only the ENGINE setting (an in-
# memory database will be used). All other backends will require a
# NAME and potentially authentication information. See the
# following section in the docs for more information:
#
# https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/unit-tests/
#
# The different databases that Django supports behave differently in certain
# situations, so it is recommended to run the test suite against as many
# database backends as possible.  You may want to create a separate settings
# file for each of the backends you test against.

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": "django",
    },
    "other": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": "django2",
    },
}

SECRET_KEY = "django_tests_secret_key"

# Use a fast hasher to speed up tests.
PASSWORD_HASHERS = [
    "django.contrib.auth.hashers.MD5PasswordHasher",
]

DEFAULT_AUTO_FIELD = "django.db.models.AutoField"

USE_TZ = False
```

I do get segfaults after a few tests:
```
./runtests.py --settings=test_postgresql --parallel=8                                                                                                                                            
Testing against Django installed in '/home/florian/sources/django.git/django' with up to 8 processes
Found 15782 test(s).
Creating test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Creating test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
System check identified no issues (17 silenced).
....Fatal Python error: Segmentation fault

Current thread 0x00007ff582c5b740 (most recent call first):
  File "/home/florian/sources/django.git/django/db/backends/base/base.py", line 364 in close
  File "/home/florian/sources/django.git/django/utils/asyncio.py", line 26 in inner
  File "/home/florian/sources/django.git/django/test/testcases.py", line 1461 in tearDownClass
  File "/usr/lib64/python3.10/unittest/suite.py", line 306 in _tearDownPreviousClass
  File "/usr/lib64/python3.10/unittest/suite.py", line 130 in run
  File "/usr/lib64/python3.10/unittest/suite.py", line 84 in __call__
  File "/home/florian/sources/django.git/django/test/runner.py", line 366 in run
  File "/home/florian/sources/django.git/django/test/runner.py", line 446 in _run_subsuite
  File "/usr/lib64/python3.10/multiprocessing/pool.py", line 125 in worker
  File "/usr/lib64/python3.10/multiprocessing/process.py", line 108 in run
  File "/usr/lib64/python3.10/multiprocessing/process.py", line 315 in _bootstrap
  File "/usr/lib64/python3.10/multiprocessing/popen_fork.py", line 71 in _launch
  File "/usr/lib64/python3.10/multiprocessing/popen_fork.py", line 19 in __init__
  File "/usr/lib64/python3.10/multiprocessing/context.py", line 277 in _Popen
  File "/usr/lib64/python3.10/multiprocessing/process.py", line 121 in start
  File "/usr/lib64/python3.10/multiprocessing/pool.py", line 326 in _repopulate_pool_static
  File "/usr/lib64/python3.10/multiprocessing/pool.py", line 303 in _repopulate_pool
  File "/usr/lib64/python3.10/multiprocessing/pool.py", line 212 in __init__
  File "/usr/lib64/python3.10/multiprocessing/context.py", line 119 in Pool
  File "/home/florian/sources/django.git/django/test/runner.py", line 497 in run
  File "/usr/lib64/python3.10/unittest/suite.py", line 84 in __call__
  File "/usr/lib64/python3.10/unittest/runner.py", line 184 in run
  File "/home/florian/sources/django.git/django/test/runner.py", line 967 in run_suite
  File "/home/florian/sources/django.git/django/test/runner.py", line 1045 in run_tests
  File "/home/florian/sources/django.git/tests/./runtests.py", line 427 in django_tests
  File "/home/florian/sources/django.git/tests/./runtests.py", line 768 in <module>

Extension modules: pywatchman.bser, markupsafe._speedups, PIL._imaging, psycopg2._psycopg, PIL._webp, yaml._yaml, numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, _cffi_backend (total: 20)
```

Now I am not 100% sure yet if this is due to my code changes or not, but Python should not crash. Then again since psycopg uses ctypes there is probably a chance of crashes? The other option is that this is a python bug but I am opening it here first and maybe we can figure out if it is psycopg or python. I can also gather core dumps, but it would be great if someone can try verifying it on another machine so I am not chasing heisenbugs.

I also do not observe this issue if I use 'psycopg[binary]', which points even more towards a psycopg ctypes or python bug. I will try to gather more details tomorrow, but if you have any ideas let me know.

### PR DESCRIPTION
**Pull request by apollo13**
The connection is passed via the optional *arg.

Fixes #300

### CODE REVIEW

File: psycopg/psycopg/pq/pq_ctypes.py [hunk 1]
```diff
@@ -81,10 +79,11 @@ def __init__(self, pgconn_ptr: impl.PGconn_struct):
         self.notice_handler: Optional[Callable[["abc.PGresult"], None]] = None
         self.notify_handler: Optional[Callable[[PGnotify], None]] = None
 
-        self._notice_receiver = impl.PQnoticeReceiver(  # type: ignore
-            partial(notice_receiver, wconn=ref(self))
+        # Keep alive for the duration of PGconn
```
Comment (dvarrazzo) [inline psycopg/psycopg/pq/pq_ctypes.py:82]:
```suggestion
        # Keep alive for the lifetime of PGconn
```

File: psycopg/psycopg/pq/pq_ctypes.py [hunk 2]
```diff
@@ -71,7 +69,7 @@ class PGconn:
         "_pgconn_ptr",
         "notice_handler",
         "notify_handler",
-        "_notice_receiver",
+        "_py_object_helper",
```
Comment (dvarrazzo) [inline psycopg/psycopg/pq/pq_ctypes.py:72]: Maybe another name? `_self_ptr`?

**Review by dvarrazzo — CHANGES_REQUESTED** [review]
I have tried faffing a bit with the ctypes object and haven't found a more streamlined way to do what you are doing now, so I think we are ok with this.

**Comment by apollo13** [comment]
I had to adjust the test a bit to use async functions because otherwise `poll.map` simply won't finish if the subprocess crashes (keeps waiting for results). The timeout of 10 seconds is rather long but should only be an issue if the error condition is triggered, so it should not happen. I verified that the test breaks when my changes are not applied.

**Comment by apollo13** [comment]
Ok, now I am confused. I'd expect the mac/windows tests to fail consistently if they cannot pickle `__main__.test`. Why do they sometimes work and sometimes not? Also this test only makes sense when multiprocessing uses fork a method, should we just adjust it to skip other systems?

**Review by dvarrazzo — APPROVED** [review]
Looks ok to me.

I've rebased your changes on the maint-3.0 branch and pushed to [run the tests](https://github.com/psycopg/psycopg/actions/runs/2942607003).

It occurred to me that you have worked on the master branch (to become 3.1), but the first tests you made were on a 3.0.x version. Could you please run a quick test on your system to verify that the changes applied work even on the [pgnotifyfixes-3.0 branch](https://github.com/psycopg/psycopg/tree/pgnotifyfixes-3.0)?

Thank you very much!

**Comment by apollo13** [comment]
I ran that single test against 3.0 and it worked fine. That said I get loads of failures in the async tests on 3.0:
```
FAILED tests/test_connection_async.py::test_auto_transaction - AttributeError: 'async_generator' object has no attribute 'pgconn'
FAILED tests/test_connection_async.py::test_autocommit_unknown - AttributeError: 'async_generator' object has no attribute 'close'
FAILED tests/test_connection_async.py::test_autocommit_inerror - AttributeError: 'async_generator' object has no attribute 'cursor'
FAILED tests/test_connection_async.py::test_transaction_param_default[isolation_level] - AttributeError: 'async_generator' object has no attribute 'isolation_level'
FAILED tests/test_connection_async.py::test_set_transaction_param_not_intrans_external[deferrable] - AttributeError: 'async_generator' object has no attribute 'set_autocommit'
FAILED tests/test_connection_async.py::test_execute_binary - AttributeError: 'async_generator' object has no attribute 'execute'
FAILED tests/test_connection_async.py::test_notice_handlers - AttributeError: 'async_generator' object has no attribute 'add_notice_handler'
FAILED tests/test_connection_async.py::test_cursor_factory - AttributeError: 'async_generator' object has no attribute 'cursor_factory'
FAILED tests/test_connection_async.py::test_set_transaction_param_block[isolation_level-True] - AttributeError: 'async_generator' object has no attribute 'set_autocommit'
FAILED tests/test_connection_async.py::test_transaction_param_default[deferrable] - AttributeError: 'async_generator' object has no attribute 'deferrable'
FAILED tests/test_connection_async.py::test_set_transaction_param_implicit[isolation_level-True] - AttributeError: 'async_generator' object has no attribute 'set_autocommit'
```
but those do happen even before my changes and are gone on 3.1. Not sure what is going on here.

**Comment by dvarrazzo** [comment]
Do you mean that [these tests](https://github.com/psycopg/psycopg/runs/8056831752?check_suite_focus=true#step:8:41) fail for you?

Please note that tests in 3.0 and 3.1 use a different pytest-asyncio version: maybe the difference comes from there?

**Comment by apollo13** [comment]
> Do you mean that [these tests](https://github.com/psycopg/psycopg/runs/8056831752?check_suite_focus=true#step:8:41) fail for you?

Not just those but most tests including `_async.py` in their filename. That said I ran the new test manually and also ran the 3.0 branch against Django. Given that the tests work in CI (handlers are tested there after all) and the commit doesn't make things worse locally I am not going to inspect why the 3.0 testsuite doesn't work for me locally (although your pstest-asyncio might be the culprit).

**Comment by dvarrazzo** [comment]
That's ok: your changes here are an improvement and I'm happy to merge them. If there is any fallout from them, we will deal with it afterwards.

By the way, [the Cython implementation](https://github.com/psycopg/psycopg/blob/c4c3fbee0258ddaae08c4f54502c66be2a9692fc/psycopg_c/psycopg_c/pq/pgconn.pyx#L41) is pretty similar what you have done here. I think that was my original intention, but I couldn't get to the bottom of it because my knowledge of ctypes was much poorer. So, I'm happy about this your implementation, thank you.

I'll force-pushing my changes here (I reworded the commit message and added a news entry) and merging to master too.

**Pull request merged**